test/s3_test.go

```diff
@@ -0,0 +1,287 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestTerraformS3Bucket(t *testing.T) {
+	t.Parallel()
+
+	// For convenience - uncomment these as well as the "os" import
+	// when doing local testing if you need to skip any sections.
+	//
+	// A common usage for this would be skipping teardown first (keeping infrastructure)
+	// and then in your next run skip the setup* and create* steps. This way you can keep testing
+	// your Go test against your infrastructure quicker. Be mindful of random-ids, as they would be updated
+	// on each run, which would make some assertions fail.
+
+	// os.Setenv("SKIP_", "true")
+	// os.Setenv("TERRATEST_REGION", "us-east-1")
+
+	// os.Setenv("SKIP_setup_options", "true")
+	// os.Setenv("SKIP_create_bucket", "true")
+	// os.Setenv("SKIP_validate_bucket", "true")
+	// os.Setenv("SKIP_create_ro_objects", "true")
+	// os.Setenv("SKIP_setup_role_options", "true")
+	// os.Setenv("SKIP_create_role", "true")
+	// os.Setenv("SKIP_validate_bucket_and_policies", "true")
+
+	// os.Setenv("SKIP_teardown", "true")
+	// os.Setenv("SKIP_teardown_role", "true")
+
+	// string to be used as body for files to be created
+	testBodyString := "test"
+
+	// define struct for each object Test Case
+	type objectTestCase struct {
+		key             string
+		encryption      string
+		expectPassRead  bool
+		expectPassWrite bool
+	}
+
+	// list of different buckets that will be created to be tested
+	var bucketTestCases = []struct {
+		testName      string
+		pathRO        []string
+		pathRW        []string
+		objTestCases  []objectTestCase
+		sleepDuration int
+	}{
+		{
+			"TestBucket1",
+			[]string{"path/to/ro-folder"},
+			[]string{"path/to/rw-folder"},
+			[]objectTestCase{
+				{
+					key:             "path/to/ro-folder/obj1",
+					encryption:      "AES256",
+					expectPassRead:  true,
+					expectPassWrite: false,
+				},
+				{
+					key:             "path/to/rw-folder/obj2",
+					encryption:      "AES256",
+					expectPassRead:  true,
+					expectPassWrite: true,
+				},
+				{
+					key:             "other/folder/obj3",
+					encryption:      "AES256",
+					expectPassRead:  false,
+					expectPassWrite: false,
+				},
+			},
+			0,
+		},
+		{
+			"TestBucket2",
+			[]string{"other/path/to/ro-folder"},
+			[]string{"other/path/to/rw-folder"},
+			[]objectTestCase{
+				{
+					key: "path/to/ro-folder/obj1",
+					// encryption:      "",
+					expectPassRead:  false,
+					expectPassWrite: false,
+				},
+				{
+					key: "path/to/rw-folder/obj2",
+					// encryption:      "",
+					expectPassRead:  false,
+					expectPassWrite: false,
+				},
+				{
+					key: "other/rw-folder/obj3",
+					// encryption:      "",
+					expectPassRead:  false,
+					expectPassWrite: false,
+				},
+			},
+			0,
+		},
+	}
+
+	awsRegion := aws.GetRandomStableRegion(t, []string{"us-east-1", "us-east-2", "us-west-1", "us-west-2"}, nil)
```

Comment:
This function will check if `TERRATEST_REGION` is set, if so, returns that instead of the random one. Comments have been added to make it less confusing.